File: scala/compass.scala
// START:constructor
class Compass {

  val directions = List("north", "east", "south", "west")
  var bearing = 0

  print("Initial bearing: ")
  println(direction)

// END:constructor
// START:direction
  def direction() = directions(bearing)

  def inform(turnDirection: String) {
    println("Turning " + turnDirection + ". Now bearing " + direction)
  }
// END:direction
  
// START:turns
  def turnRight() {
    bearing = (bearing + 1) % directions.size
    inform("right")
  }

  def turnLeft() {
    bearing = (bearing + (directions.size - 1)) % directions.size
    inform("left")
  }
// END:turns
}

val myCompass = new Compass

myCompass.turnRight
myCompass.turnRight

myCompass.turnLeft
myCompass.turnLeft
myCompass.turnLeft

File: clojure/compass.clj
; START:protocol
(defprotocol Compass
  (direction [c])
  (left [c])
  (right [c]))
; END:protocol

; START:directions
(def directions [:north :east :south :west])
; END:directions

; START:turn
(defn turn
  [base amount]
  (rem (+ base amount) (count directions)))
; END:turn

; START:defrecord
(defrecord SimpleCompass [bearing]
  Compass
; END:defrecord
; START:directionfunction
  (direction [_] (directions bearing))
; END:directionfunction
; START:leftandright
  (left [_] (SimpleCompass. (turn bearing 3)))
  (right [_] (SimpleCompass. (turn bearing 1)))
; END:leftandright
; START:object
  Object
  (toString [this] (str "[" (direction this) "]")))
; END:object

